File: kotlin/examples/src/main/kotlin/benchmarkUtils.kt
package gr.elaevents.buffela.examples

import kotlin.system.measureTimeMillis

private const val warmupReps = 100_000
private const val measurementReps = 10_000_000

fun benchmark(label: String, action: () -> Unit) {
    for (i in 0..warmupReps) action()

    val totalTimeMs = measureTimeMillis { for (i in 0..measurementReps) action() }
    val repTimeMs = (totalTimeMs.toDouble() / measurementReps).toBigDecimal()
    println("$label took ${repTimeMs.toPlainString()} ms on average")
}

File: js/examples/utils/benchmarkUtils.js
const { performance } = require('node:perf_hooks');

const warmupReps = 100_000
const measurementReps = 1_000_000

function benchmark(label, action) {
    for (let i = 0; i < warmupReps; i++) action()

    const startTimeMs = performance.now()
    for (let i = 0; i < measurementReps; i++) action()
    const endTimeMs = performance.now()

    const repTimeMs = (endTimeMs - startTimeMs) / measurementReps
    console.log(`${label} took ${repTimeMs}ms on average`)
}

module.exports = { benchmark }